src/main/java/com/google/firebase/internal/RetryConfig.java

```diff
@@ -0,0 +1,160 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Configures when and how HTTP requests should be retried.
+ */
+public final class RetryConfig {
+
+  private static final int INITIAL_INTERVAL_MILLIS = 500;
+
+  private final List<Integer> retryStatusCodes;
+  private final int maxRetries;
+  private final Sleeper sleeper;
+  private final ExponentialBackOff.Builder backOffBuilder;
+
+  private RetryConfig(Builder builder) {
+    if (builder.retryStatusCodes != null) {
+      this.retryStatusCodes = ImmutableList.copyOf(builder.retryStatusCodes);
+    } else {
+      this.retryStatusCodes = ImmutableList.of();
+    }
+
+    checkArgument(builder.maxRetries >= 0, "maxRetries must not be negative");
+    this.maxRetries = builder.maxRetries;
+    this.sleeper = checkNotNull(builder.sleeper);
+    this.backOffBuilder = new ExponentialBackOff.Builder()
+        .setInitialIntervalMillis(INITIAL_INTERVAL_MILLIS)
+        .setMaxIntervalMillis(builder.maxIntervalMillis)
+        .setMultiplier(builder.backOffMultiplier)
+        .setRandomizationFactor(0);
+
+    // Force validation of arguments by building the BackOff object
+    this.backOffBuilder.build();
+  }
+
+  List<Integer> getRetryStatusCodes() {
+    return retryStatusCodes;
+  }
+
+  int getMaxRetries() {
+    return maxRetries;
+  }
+
+  int getMaxIntervalMillis() {
+    return backOffBuilder.getMaxIntervalMillis();
+  }
+
+  double getBackOffMultiplier() {
+    return backOffBuilder.getMultiplier();
+  }
+
+  Sleeper getSleeper() {
+    return sleeper;
+  }
+
+  BackOff newBackOff() {
+    return backOffBuilder.build();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+
+    private List<Integer> retryStatusCodes;
+    private int maxRetries;
+    private int maxIntervalMillis = (int) TimeUnit.MINUTES.toMillis(2);
+    private double backOffMultiplier = 2.0;
+    private Sleeper sleeper = Sleeper.DEFAULT;
+
+    private Builder() { }
+
+    /**
+     * Sets a list of HTTP status codes that should be retried. If null or empty, HTTP requests
+     * will not be retried as long as they result in some HTTP response message. I/O errors
+     * will still be retried.
+     *
+     * @param retryStatusCodes A list of status codes.
+     * @return This builder.
+     */
+    public Builder setRetryStatusCodes(List<Integer> retryStatusCodes) {
```

Comment:
Do you foresee that these status codes will be different for each service?

Reply:
By default we are going to retry on 503 and 500 errors (as recommended at https://cloud.google.com/apis/design/errors#error_retries). But after talking to a few teams about this, we have also decided to make these customizable for individual services. For example, if Auth doesn't want to retry on 500 errors, they should be able to.